Receiving an error when saving exercise

## Summary

Receiving an error when saving an exercise. No further changes to the exercise are possible.

Note: The exercise contains images in the question as well as in answers.

## Category

Select one: BUG 

## Usage Details

*Please give details about how you were using Studio.*

 - Browser: Firefox 64.0 64-bit and Google Chrome Version 71.0.3578.98 (Official Build) (64-bit) on Windows
 - Channel: ID - d647859ea17242a382a9f0a001520e9a
 - Topic/Content Item: An exercise with title - "Parts of a computer: Input output devices - Exercise [01-DL-01-WS]"


## Traceback or relevant snippet from browser console

```
AJAX Request Error: error
raven.js:58 Error data: {"type":"PUT","url":"/api/assessmentitem","data":"[{\"type\":\"single_selection\",\"question\":\"One day, Tina and Sita were drawing on a piece of paper. They finished drawing and went to show it to their mother. Mother said, “Very nice! Let’s show this to grandpa and grandma too!”  \\n\\n“How?” asked Tina.  \\n\\n“Well, let’s use the computer!” said mother.  \\n\\n“It’s that black coloured thing which father uses for work! Right, mother?” said Sita.  \\n\\n“Very good, Sita! Yes, it is! Let’s go see it”, said mother. They went into father’s room to see the computer.  \\n\\n“Wow! This computer looks so nice!” said Tina, running to the computer. “Be careful!”, said mother.  \\n\\n“This computer has so many parts… why does the computer have so many parts?” asked Tina.  \\n\\n“That is a great question!” said mother. “Every part of the computer is important.”  \\n\\nMother explained the parts of a computer to Sita and Tina..  \\n\\n![cpu image](${☣ CONTENTSTORAGE}/cd5aeac3e24d2f91e3f10c6e3a858a2d.jpg) CPU ![monitor image](${☣ CONTENTSTORAGE}/d788660b46174b2b15fd7840cbf039ee.jpg) Monitor ![mouse image](${☣ CONTENTSTORAGE}/91fc263aa8ad1c30beadaea893bf55f9.jpg) Mouse  \\n\\n![keyboard image](${☣ CONTENTSTORAGE}/430bbfcf09e4b50c5098ad646da2723c.jpg) Keyboard ![speakers image](${☣ CONTENTSTORAGE}/fae122c1a1ae633d9cab41bd11bb8e9c.png) Speakers\\n\\n##   \\n![Webcam](${☣ CONTENTSTORAGE}/66ff2197ede027c74f671c16329d769c.png) Webcam\\n\\nI want to **point to the apple** on the monitor. What will I use? ![monitor_with_apple image](${☣ CONTENTSTORAGE}/af5b4720579cd2a3173d10f11988b6a6.png)\",\"answers\":\"[{\\\"order\\\":1,\\\"answer\\\":\\\"![keyboard image](${☣ CONTENTSTORAGE}/430bbfcf09e4b50c5098ad646da2723c.jpg)\\\",\\\"correct\\\":false,\\\"files\\\":[\\\"0a2670d19ee3402fab2135376f3bb79d\\\",\\\"b34345dbf3e441bf94e1d3886726590d\\\"]},{\\\"order\\\":2,\\\"answer\\\":\\\"![mouse image](${☣ CONTENTSTORAGE}/91fc263aa8ad1c30beadaea893bf55f9.jpg)\\\",\\\"correct\\\":true,\\\"files\\\":[\\\"8408fed057a844a3b1029a55c1f95d57\\\"]},{\\\"order\\\":3,\\\"answer\\\":\\\"![speakers image](${☣ CONTENTSTORAGE}/fae122c1a1ae633d9cab41bd11bb8e9c.png)\\\",\\\"correct\\\":false,\\\"files\\\":[\\\"c94d5ec992b047ff8640f9743f49ebef\\\"]}]\",\"hints\":\"[]\",\"files\":[\"0a2670d19ee3402fab2135376f3bb79d\",\"b34345dbf3e441bf94e1d3886726590d\",\"8408fed057a844a3b1029a55c1f95d57\",\"c94d5ec992b047ff8640f9743f49ebef\",\"62216b669c4641ff913191261c7a97c8\",\"b47ae562321e43c28a01c5edb2a8067c\",\"30830310a47d49fdb8c1cea4b3bd3c91\",\"e4cde38225824a9f8f974581beeb7011\",\"09f93700c9904d778a170edc814a22a4\",\"c739baa355aa46ee85a2fc0e98072087\",\"bf64a79b1fec4ef69e2f6f0007f011c4\"],\"id\":29304420,\"contentnode\":\"83dc4cb593a54175acc048a02f0435cf\",\"assessment_id\":\"bdf9d15a516140fa968dd3054ec33674\",\"raw_data\":\"\",\"order\":1,\"source_url\":null,\"randomize\":true,\"deleted\":false},{\"type\":\"single_selection\",\"question\":\"I want to **type a letter** to send to my friend. What will I use?\",\"answers\":\"[{\\\"order\\\":1,\\\"answer\\\":\\\"![keyboard image](${☣ CONTENTSTORAGE}/430bbfcf09e4b50c5098ad646da2723c.jpg)\\\",\\\"correct\\\":true,\\\"files\\\":[\\\"29caf6c67c2443d59f16bc3a8e0f1f63\\\"]},{\\\"order\\\":2,\\\"answer\\\":\\\"![mouse image](${☣ CONTENTSTORAGE}/91fc263aa8ad1c30beadaea893bf55f9.jpg)\\\",\\\"correct\\\":false,\\\"files\\\":[\\\"f9665fd49b8b47c1905c9e197ed36514\\\"]},{\\\"order\\\":3,\\\"answer\\\":\\\"![speakers image](${☣ CONTENTSTORAGE}/fae122c1a1ae633d9cab41bd11bb8e9c.png)\\\",\\\"correct\\\":false,\\\"files\\\":[\\\"703a7f39ca564d778bd2ef29380ad586\\\"]}]\",\"hints\":\"[]\",\"files\":[\"29caf6c67c2443d59f16bc3a8e0f1f63\",\"f9665fd49b8b47c1905c9e197ed36514\",\"703a7f39ca564d778bd2ef29380ad586\"],\"id\":29463805,\"contentnode\":\"83dc4cb593a54175acc048a02f0435cf\",\"assessment_id\":\"1994fd41499e464f84b94589f73cb575\",\"raw_data\":\"\",\"order\":2,\"source_url\":null,\"randomize\":true,\"deleted\":false},{\"type\":\"single_selection\",\"question\":\"I want to **listen to a song**. What will I use?\",\"answers\":\"[{\\\"order\\\":1,\\\"answer\\\":\\\"![keyboard image](${☣ CONTENTSTORAGE}/430bbfcf09e4b50c5098ad646da2723c.jpg)\\\",\\\"correct\\\":true,\\\"files\\\":[\\\"d6514f98d18e4ed4b3b904337eb5da50\\\"]},{\\\"order\\\":2,\\\"answer\\\":\\\"![speakers image](${☣ CONTENTSTORAGE}/fae122c1a1ae633d9cab41bd11bb8e9c.png)\\\",\\\"correct\\\":false,\\\"files\\\":[\\\"3dc6907bb7b540bea5c761a60564e354\\\"]},{\\\"order\\\":3,\\\"answer\\\":\\\"![mouse image](${☣ CONTENTSTORAGE}/91fc263aa8ad1c30beadaea893bf55f9.jpg)\\\",\\\"correct\\\":false,\\\"files\\\":[\\\"d5af792959bb4d5c96f306ba8ebd9abf\\\"]}]\",\"hints\":\"[]\",\"files\":[\"d6514f98d18e4ed4b3b904337eb5da50\",\"3dc6907bb7b540bea5c761a60564e354\",\"d5af792959bb4d5c96f306ba8ebd9abf\"],\"id\":29463806,\"contentnode\":\"83dc4cb593a54175acc048a02f0435cf\",\"assessment_id\":\"6e8608f3c62d46ce92a1c9ba17a96941\",\"raw_data\":\"\",\"order\":3,\"source_url\":null,\"randomize\":true,\"deleted\":false}]","status":400,"error":"error","response":"[{\"files\":[\"Invalid pk \\\"0a2670d19ee3402fab2135376f3bb79d\\\" - object does not exist.\"]},{},{}]"}
```

## How to reproduce

1. Edit the exercise
2. Attempt saving it.
3. A pop-up response is received - 
```
[{"files":["Invalid pk \"0a2670d19ee3402fab2135376f3bb79d\" - object does not exist."]},{},{}]
```
4. No further edit/changes to the exercise is possible.

## Screenshots

![exercise_error_on_editing](https://user-images.githubusercontent.com/2577060/50266190-bd347980-0448-11e9-992c-97e67db683e0.gif)

--- a/contentcuration/contentcuration/serializers.py
+++ b/contentcuration/contentcuration/serializers.py
@@ -1,4 +1,5 @@
 import json
+import re
 from collections import OrderedDict
 
 from django.conf import settings
@@ -267,38 +268,43 @@ class TagSerializer(serializers.ModelSerializer):
         fields = ('tag_name', 'channel', 'id')
 
 
+exercise_image_checksum_regex = re.compile("\!\[[^]]*\]\(\$" + exercises.IMG_PLACEHOLDER + "/([a-f0-9]{32})\.[\w]+\)")
+
+
 class AssessmentListSerializer(serializers.ListSerializer):
 
-    def update(self, instance, validated_data):
+    def update(self, queryset, validated_data):
         ret = []
 
+        validated_data_by_id = {
+            i.pop('id'): i
+            for i in validated_data
+        }
+
         with transaction.atomic():
-            for item in validated_data:
-                files = item.pop('files', [])  # Remove here to avoid problems with setting attributes
-
-                # Handle existing items
-                if 'id' in item:
-                    aitem, is_new = AssessmentItem.objects.get_or_create(pk=item['id'])
-                    if item['deleted']:
-                        aitem.delete()
-                        continue
-                    else:
-                        # Set attributes for assessment item
-                        for attr, value in item.items():
-                            setattr(aitem, attr, value)
-                        aitem.save()
-                else:
-                    # Create item
-                    aitem = AssessmentItem.objects.create(**item)
-
-                for f in files:
-                    if f.checksum in str(aitem.__dict__):
-                        if f.assessment_item_id != aitem.pk:
-                            f.assessment_item = aitem
-                            f.save()
-                    else:
-                        f.delete()
-
+            # only handle existing items
+            aitems = AssessmentItem.objects.filter(id__in=validated_data_by_id.keys())
+            if len(validated_data) != aitems.count():
+                raise ValidationError('Could not find all objects to update')
+            for aitem in aitems:
+                item = validated_data_by_id.get(aitem.id)
+                # Defer to child serializer for update operations
+                self.child.update(aitem, item)
+                # Get unique checksums in the assessment item text fields markdown
+                # Coerce to a string, for Python 2, as the stored data is in unicode, and otherwise
+                # the unicode char in the placeholder will not match
+                checksums = set(exercise_image_checksum_regex.findall(str(aitem.question + aitem.answers + aitem.hints)))
+                # delete any files that are no longer being used
+                aitem.files.exclude(checksum__in=checksums).delete()
+                if len(checksums) != aitem.files.count():
+                    # We have some checksums that do not have files associated
+                    no_files = checksums - set(aitem.files.values_list('checksum', flat=True))
+                    for checksum in no_files:
+                        # Find an existing file object for this file
+                        file = File.objects.filter(checksum=checksum).first()
+                        file.id = None
+                        file.assessment_item = aitem
+                        file.save()
                 ret.append(aitem)
 
         return ret
@@ -310,7 +316,7 @@ class AssessmentItemSerializer(BulkSerializerMixin, serializers.ModelSerializer)
 
     class Meta:
         model = AssessmentItem
-        fields = ('id', 'question', 'files', 'type', 'answers', 'contentnode', 'assessment_id',
+        fields = ('id', 'question', 'type', 'answers', 'contentnode', 'assessment_id',
                   'hints', 'raw_data', 'order', 'source_url', 'randomize', 'deleted')
         list_serializer_class = AssessmentListSerializer
 

--- a/contentcuration/contentcuration/static/js/edit_channel/exercise_creation/views.js
+++ b/contentcuration/contentcuration/static/js/edit_channel/exercise_creation/views.js
@@ -215,6 +215,7 @@ var UploadImage = function (context) {
         click: function () {
             var view = new ImageUploader.ImageUploadView({
                 callback: context.options.callbacks.onCustomImageUpload,
+                assessmentItemId: context.options.assessmentItemId,
                 preset_id: 'exercise_image'
             });
         }
@@ -318,6 +319,7 @@ var EditorView = BaseViews.BaseView.extend({
         this.edit_key = options.edit_key;
         this.editing = false;
         this.numbersOnly = options.numbersOnly || false;
+        this.assessmentItem = options.assessmentItem;
         this.render();
         this.markdown = this.model.get(this.edit_key);
         this.listenTo(this.model, "change:" + this.edit_key, this.render);
@@ -411,6 +413,7 @@ var EditorView = BaseViews.BaseView.extend({
             shortcuts: false,
             selector: this.cid,
             disableDragAndDrop: true,
+            assessmentItemId: this.assessmentItem.id,
             callbacks: {
                 onChange: _.debounce(this.save, 1),
                 onPaste: this.paste_content,
@@ -772,7 +775,12 @@ var ExerciseEditableItemView =  BaseViews.BaseListEditableItemView.extend({
     /*********** EDITOR METHODS ***********/
     render_editor: function(){
         if (!this.editor_view) {
-            this.editor_view = new EditorView({model: this.model, edit_key: this.content_field, numbersOnly:this.numbers_only()});
+            this.editor_view = new EditorView({
+                model: this.model,
+                edit_key: this.content_field,
+                numbersOnly:this.numbers_only(),
+                assessmentItem: this.assessmentItem
+            });
         }
         this.$(this.editor_el).html(this.editor_view.el);
         this.listenTo(this.model, "change:" + this.content_field, this.propagate_changes)
@@ -870,6 +878,19 @@ var ExerciseView = ExerciseEditableListView.extend({
         'change #randomize_question_order': 'set_random',
     },
 
+    /*********** EDITING METHODS ***********/
+    add_item: function() {
+        if(!this.$(this.additem_el).hasClass('disabled')){
+            this.$(this.default_item).css('display', 'none');
+            this.close_all_editors();
+            const model = new Models.AssessmentItemModel(this.get_default_attributes());
+            model.save().then(() => {
+                this.collection.add(model);
+                this.propagate_changes();
+            });
+        }
+    },
+
     /*********** LOADING METHODS ***********/
     render: function() {
         this.$el.html(this.template({
@@ -879,7 +900,7 @@ var ExerciseView = ExerciseEditableListView.extend({
         }, {
             data: this.get_intl_data()
         }));
-        this.load_content(this.collection.where({'deleted': false}), (this.allow_edit)? this.get_translation("add_question_prompt") : this.get_translation("no_questions_found"));
+        this.load_content(this.collection, (this.allow_edit)? this.get_translation("add_question_prompt") : this.get_translation("no_questions_found"));
     },
     create_new_view:function(model){
         var new_exercise_item = null;
@@ -934,6 +955,7 @@ var AssessmentItemDisplayView = ExerciseEditableItemView.extend({
 
     initialize: function(options) {
         _.bindAll(this, "update_hints", "show_hints");
+        this.assessmentItem = this.model;
         this.render();
     },
     events: {
@@ -957,7 +979,7 @@ var AssessmentItemDisplayView = ExerciseEditableItemView.extend({
                 this.answer_editor = new AssessmentItemAnswerListView({
                     collection: this.model.get("answers"),
                     container:this,
-                    assessment_item: this.model,
+                    assessmentItem: this.model,
                     isdisplay:this.isdisplay
                 });
             }
@@ -969,7 +991,7 @@ var AssessmentItemDisplayView = ExerciseEditableItemView.extend({
             this.hint_editor = new HintModalView({
                 collection: this.model.get("hints"),
                 container: this,
-                assessment_item: this.model,
+                assessmentItem: this.model,
                 model: this.model,
                 onupdate: this.update_hints,
                 isdisplay: this.isdisplay
@@ -997,6 +1019,7 @@ var AssessmentItemView = AssessmentItemDisplayView.extend({
         this.onchange = options.onchange;
         this.question = this.model.get('question');
         this.containing_list_view = options.containing_list_view;
+        this.assessmentItem = this.model;
         this.init_undo_redo();
         this.render();
         this.set_toolbar_closed();
@@ -1023,11 +1046,12 @@ var AssessmentItemView = AssessmentItemDisplayView.extend({
         dialog.dialog(this.get_translation("deleting_question"), this.get_translation("deleting_question_text"), {
             [this.get_translation("cancel")]:function(){},
             [this.get_translation("delete")]: function(){
-                self.model.set('deleted', true);
-                self.propagate_changes();
-                self.containing_list_view.views.splice(self, 1);
-                self.containing_list_view.handle_if_empty();
-                self.remove();
+                self.model.destroy().then(() => {
+                    self.propagate_changes();
+                    self.containing_list_view.views.splice(self, 1);
+                    self.containing_list_view.handle_if_empty();
+                    self.remove();
+                });
             },
         }, function(){});
     },
@@ -1258,14 +1282,14 @@ var AssessmentItemAnswerListView = ExerciseEditableListView.extend({
         return {
             order: this.get_next_order(),
             answer: "",
-            correct: this.assessment_item.get('type') === "input_question"
+            correct: this.assessmentItem.get('type') === "input_question"
         };
     },
     initialize: function(options) {
         _.bindAll(this, "render", "add_item", "add_item_view");
         this.bind_edit_functions();
         this.views = [];
-        this.assessment_item = options.assessment_item;
+        this.assessmentItem = options.assessmentItem;
         this.isdisplay = options.isdisplay;
         this.container = options.container;
         this.render();
@@ -1274,9 +1298,9 @@ var AssessmentItemAnswerListView = ExerciseEditableListView.extend({
     },
     render: function() {
         this.$el.html(this.template({
-            input_answer: this.assessment_item.get("type") === "input_question",
+            input_answer: this.assessmentItem.get("type") === "input_question",
             isdisplay: this.isdisplay,
-            true_false: this.assessment_item.get("type") === "true_false"
+            true_false: this.assessmentItem.get("type") === "true_false"
         }, {
             data: this.get_intl_data()
         }));
@@ -1287,7 +1311,7 @@ var AssessmentItemAnswerListView = ExerciseEditableListView.extend({
         var view = new AssessmentItemAnswerView({
             model: model,
             containing_list_view:this,
-            assessment_item: this.assessment_item,
+            assessmentItem: this.assessmentItem,
             isdisplay:this.isdisplay
         });
         this.views.push(view);
@@ -1317,7 +1341,7 @@ var AssessmentItemAnswerView = ExerciseEditableItemView.extend({
         _.bindAll(this, "render", "set_editor", "set_open", "toggle", "toggle_correct");
         this.open = options.open || false;
         this.containing_list_view = options.containing_list_view;
-        this.assessment_item = options.assessment_item;
+        this.assessmentItem = options.assessmentItem;
         this.isdisplay = options.isdisplay;
         this.render();
     },
@@ -1333,9 +1357,9 @@ var AssessmentItemAnswerView = ExerciseEditableItemView.extend({
             answer: this.model.toJSON(),
             input_answer: this.numbers_only(),
             single_selection: this.is_single_correct(),
-            groupName: this.assessment_item.cid,
+            groupName: this.assessmentItem.cid,
             allow_edit: !this.isdisplay,
-            is_true_false: this.assessment_item.get("type") === "true_false",
+            is_true_false: this.assessmentItem.get("type") === "true_false",
             allow_toggle: !this.isdisplay
         }, {
             data: this.get_intl_data()
@@ -1343,9 +1367,9 @@ var AssessmentItemAnswerView = ExerciseEditableItemView.extend({
         this.render_editor();
         _.defer(this.set_editor);
     },
-    numbers_only: function() {return this.assessment_item.get('type') === 'input_question';},
+    numbers_only: function() {return this.assessmentItem.get('type') === 'input_question';},
     is_single_correct: function(){
-        return this.assessment_item.get("type") === "single_selection" || this.assessment_item.get("type") === "true_false";
+        return this.assessmentItem.get("type") === "single_selection" || this.assessmentItem.get("type") === "true_false";
     },
     toggle_correct: function(event) {
         event.stopPropagation();
@@ -1373,6 +1397,7 @@ var HintQuestionDisplayView = BaseViews.BaseView.extend({
     $trs: MESSAGES,
 
     initialize: function(options) {
+        this.assessmentItem = options.assessmentItem;
         this.render();
     },
     render: function() {
@@ -1381,6 +1406,7 @@ var HintQuestionDisplayView = BaseViews.BaseView.extend({
         }));
         var editor_view = new EditorView({
             model: this.model,
+            assessmentItemId: this.assessmentItem.id,
             edit_key: "question",
             el: this.$(".question")
         });
@@ -1397,7 +1423,7 @@ var HintModalView = BaseViews.BaseModalView.extend({
 
     initialize: function(options) {
         _.bindAll(this, "closing_hints", "show", "closed_hints", "init_focus", "loop_focus");
-        this.assessment_item = options.assessment_item;
+        this.assessmentItem = options.assessmentItem;
         this.isdisplay = options.isdisplay;
         this.onupdate = options.onupdate;
         this.container = options.container;
@@ -1412,7 +1438,8 @@ var HintModalView = BaseViews.BaseModalView.extend({
         }));
         if(!this.isdisplay){
             var question_preview = new HintQuestionDisplayView({
-                model: this.assessment_item,
+                assessmentItem: this.assessmentItem,
+                model: this.model,
                 el: this.$(".question_preview")
             });
         }
@@ -1420,7 +1447,7 @@ var HintModalView = BaseViews.BaseModalView.extend({
         this.hint_editor = new AssessmentItemHintListView({
             collection: this.model.get("hints"),
             container: this.container,
-            assessment_item: this.model,
+            assessmentItem: this.model,
             model: this.model,
             onupdate: this.onupdate,
             isdisplay: this.isdisplay,
@@ -1468,7 +1495,7 @@ var AssessmentItemHintListView = ExerciseEditableListView.extend({
     initialize: function(options) {
         _.bindAll(this, "render", "add_item", "add_item_view");
         this.bind_edit_functions();
-        this.assessment_item = options.assessment_item;
+        this.assessmentItem = options.assessmentItem;
         this.isdisplay = options.isdisplay;
         this.modal_view = options.modal_view;
         this.render();
@@ -1492,7 +1519,7 @@ var AssessmentItemHintListView = ExerciseEditableListView.extend({
         var view = new AssessmentItemHintView({
             model: model,
             containing_list_view:this,
-            assessment_item: this.assessment_item,
+            assessmentItem: this.assessmentItem,
             isdisplay: this.isdisplay
         });
         this.views.push(view);
@@ -1521,7 +1548,7 @@ var AssessmentItemHintView = ExerciseEditableItemView.extend({
         _.bindAll(this, "render", "set_editor", "set_open", "toggle");
         this.open = options.open || false;
         this.containing_list_view = options.containing_list_view;
-        this.assessment_item = options.assessment_item;
+        this.assessmentItem = options.assessmentItem;
         this.isdisplay = options.isdisplay;
         this.render();
         this.set_toolbar_closed();

--- a/contentcuration/contentcuration/static/js/edit_channel/image/views.js
+++ b/contentcuration/contentcuration/static/js/edit_channel/image/views.js
@@ -349,6 +349,7 @@ var ImageUploadView = BaseViews.BaseModalView.extend({
         this.callback = options.callback;
         this.file = this.alt_text = null;
         this.preset_id = options.preset_id;
+        this.assessmentItemId = options.assessmentItemId;
         this.render();
     },
 
@@ -390,7 +391,10 @@ var ImageUploadView = BaseViews.BaseModalView.extend({
                 thumbnailHeight:null,
                 previewTemplate:this.dropzone_template(null, { data: this.get_intl_data() }),
                 previewsContainer: "#dropzone",
-                headers: {"X-CSRFToken": get_cookie("csrftoken")}
+                headers: {"X-CSRFToken": get_cookie("csrftoken")},
+                params: {
+                  assessment_item: this.assessmentItemId,
+                },
             });
             this.dropzone.on("success", this.file_uploaded);
             this.dropzone.on("addedfile", this.file_added);

--- a/contentcuration/contentcuration/tests/test_fileendpoints.py
+++ b/contentcuration/contentcuration/tests/test_fileendpoints.py
@@ -1,163 +1,25 @@
-import tempfile
+from io import BytesIO
 
-import pytest
-from django.conf import settings
-from mixer.backend.django import mixer
+from base import StudioTestCase
+from django.core.urlresolvers import reverse
 
-pytestmark = pytest.mark.django_db
-
-
-@pytest.fixture
-def video():
-    return mixer.blend('contentcuration.ContentKind', kind='video')
+from contentcuration.models import AssessmentItem
+from contentcuration.models import File
 
 
-@pytest.fixture
-def preset_video(video):
-    return mixer.blend('contentcuration.FormatPreset', id='mp4', kind=video)
+class ExerciseImageUploadTestCase(StudioTestCase):
 
+    def test_post_exercise_image(self):
+        File.objects.all().delete()
+        img = BytesIO(b'mybinarydata')
+        img.name = 'myimage.jpg'
+        self.client.post(reverse('exercise_image_upload'), {0: img})
+        self.assertEqual(File.objects.all().count(), 1)
 
-@pytest.fixture
-def fileformat_mp4():
-    return mixer.blend('contentcuration.FileFormat', extension='mp4', mimetype='application/video')
-
-
-@pytest.yield_fixture
-def fileobj_video(preset_video, fileformat_mp4):
-    randomfilebytes = "4"
-
-    with tempfile.NamedTemporaryFile(dir=settings.STORAGE_ROOT, mode='w+t', delete=False) as f:
-        filename = f.name
-        f.write(randomfilebytes)
-        f.flush()
-        db_file_obj = mixer.blend('contentcuration.File', file_format=fileformat_mp4, preset=preset_video, file_on_disk=filename)
-
-        yield db_file_obj
-
-
-@pytest.fixture
-def audio():
-    return mixer.blend('contentcuration.ContentKind', kind='audio')
-
-
-@pytest.fixture
-def preset_audio(audio):
-    return mixer.blend('contentcuration.FormatPreset', id='mp3', kind=audio)
-
-
-@pytest.fixture
-def fileformat_mp3():
-    return mixer.blend('contentcuration.FileFormat', extension='mp3', mimetype='application/audio')
-
-
-@pytest.yield_fixture
-def fileobj_audio(preset_audio, fileformat_mp3):
-    randomfilebytes = "4"
-
-    with tempfile.NamedTemporaryFile(dir=settings.STORAGE_ROOT, mode='w+t', delete=False) as f:
-        filename = f.name
-        f.write(randomfilebytes)
-        f.flush()
-        db_file_obj = mixer.blend('contentcuration.File', file_format=fileformat_mp3, preset=preset_audio, file_on_disk=filename)
-
-        yield db_file_obj
-
-
-@pytest.fixture
-def exercise():
-    return mixer.blend('contentcuration.ContentKind', kind='exercise')
-
-
-@pytest.fixture
-def preset_exercise(exercise):
-    return mixer.blend('contentcuration.FormatPreset', id='perseus', kind=exercise)
-
-
-@pytest.fixture
-def fileformat_perseus():
-    return mixer.blend('contentcuration.FileFormat', extension='perseus', mimetype='application/perseus')
-
-
-@pytest.yield_fixture
-def fileobj_exercise(preset_exercise, fileformat_perseus):
-    randomfilebytes = "4"
-
-    with tempfile.NamedTemporaryFile(dir=settings.STORAGE_ROOT, mode='w+t', delete=False) as f:
-        filename = f.name
-        f.write(randomfilebytes)
-        f.flush()
-        db_file_obj = mixer.blend('contentcuration.File', file_format=fileformat_perseus, preset=preset_exercise, file_on_disk=filename)
-
-        yield db_file_obj
-
-
-@pytest.fixture
-def document():
-    return mixer.blend('contentcuration.ContentKind', kind='document')
-
-
-@pytest.fixture
-def preset_document(document):
-    return mixer.blend('contentcuration.FormatPreset', id='pdf', kind=document)
-
-
-@pytest.fixture
-def fileformat_pdf():
-    return mixer.blend('contentcuration.FileFormat', extension='pdf', mimetype='application/pdf')
-
-
-@pytest.yield_fixture
-def fileobj_document(preset_document, fileformat_pdf):
-    randomfilebytes = "4"
-
-    with tempfile.NamedTemporaryFile(dir=settings.STORAGE_ROOT, mode='w+t', delete=False) as f:
-        filename = f.name
-        f.write(randomfilebytes)
-        f.flush()
-        db_file_obj = mixer.blend('contentcuration.File', file_format=fileformat_pdf, preset=preset_document, file_on_disk=filename)
-
-        yield db_file_obj
-
-
-@pytest.fixture
-def fileobj_id1():
-    return 'notarealid.pdf'
-
-
-@pytest.fixture
-def fileobj_id2():
-    return 'notarealid.mp3'
-
-
-@pytest.fixture
-def fileobj_id3():
-    return 'notarealid.mp4'
-
-
-@pytest.fixture
-def fileobj_id4():
-    return 'notarealid.perseus'
-
-
-@pytest.fixture
-def file_list(fileobj_video, fileobj_audio, fileobj_document, fileobj_exercise, fileobj_id1, fileobj_id2, fileobj_id3, fileobj_id4):
-    return [
-        fileobj_video.__str__(),
-        fileobj_audio.__str__(),
-        fileobj_document.__str__(),
-        fileobj_exercise.__str__(),
-        fileobj_id1,
-        fileobj_id2,
-        fileobj_id3,
-        fileobj_id4,
-    ]
-
-
-@pytest.fixture
-def file_diff(fileobj_id1, fileobj_id2, fileobj_id3, fileobj_id4):
-    return [
-        fileobj_id1,
-        fileobj_id2,
-        fileobj_id3,
-        fileobj_id4,
-    ]
+    def test_post_exercise_image_assessment_item_id(self):
+        File.objects.all().delete()
+        item = AssessmentItem.objects.create()
+        img = BytesIO(b'mybinarydata')
+        img.name = 'myimage.jpg'
+        self.client.post(reverse('exercise_image_upload'), {0: img, 'assessment_item_id': item.id})
+        self.assertEqual(File.objects.get().assessment_item, item)

--- a/contentcuration/contentcuration/tests/test_rest_framework.py
+++ b/contentcuration/contentcuration/tests/test_rest_framework.py
@@ -4,6 +4,12 @@ import pytest
 from base import BaseAPITestCase
 from django.core.urlresolvers import reverse_lazy
 
+from le_utils.constants import content_kinds
+from le_utils.constants import exercises
+
+from contentcuration.models import AssessmentItem
+from contentcuration.models import ContentNode
+from contentcuration.models import File
 from contentcuration.models import User
 
 pytestmark = pytest.mark.django_db
@@ -36,3 +42,84 @@ class ChannelTestCase(BaseAPITestCase):
         response = self.put(url, {'version': original_version + 1, 'content_defaults': {}, 'pending_editors': []})
         self.channel.refresh_from_db()
         self.assertEqual(original_version, self.channel.version)
+
+
+class AssessmentItemTestCase(BaseAPITestCase):
+
+    def test_bulk_update(self):
+        exercise = ContentNode.objects.filter(kind=content_kinds.EXERCISE).first()
+        item1 = AssessmentItem.objects.create(contentnode=exercise)
+        item2 = AssessmentItem.objects.create(contentnode=exercise)
+        item3 = AssessmentItem.objects.create(contentnode=exercise)
+        item1dict = {}
+        item2dict = {}
+        item3dict = {}
+        for field in AssessmentItem._meta.fields:
+            attname = field.attname
+            set_attname = attname
+            if attname == 'contentnode_id':
+                set_attname = 'contentnode'
+            item1dict[set_attname] = getattr(item1, attname)
+            item2dict[set_attname] = getattr(item2, attname)
+            item3dict[set_attname] = getattr(item3, attname)
+        item1dict["question"] = "test"
+        item2dict["type"] = "test"
+        self.client.put(reverse_lazy('assessmentitem-list'), json.dumps([item1dict, item2dict, item3dict]), content_type="application/json")
+        item1.refresh_from_db()
+        self.assertEqual(item1.question, 'test')
+        item2.refresh_from_db()
+        self.assertEqual(item2.type, 'test')
+        item3.refresh_from_db()
+        self.assertEqual(item3.question, item3dict['question'])
+
+    def test_bulk_update_non_existent_item(self):
+        exercise = ContentNode.objects.filter(kind=content_kinds.EXERCISE).first()
+        item1 = AssessmentItem.objects.create(contentnode=exercise)
+        item1dict = {}
+        item2dict = {}
+        item3dict = {}
+        for field in AssessmentItem._meta.fields:
+            attname = field.attname
+            set_attname = attname
+            if attname == 'contentnode_id':
+                set_attname = 'contentnode'
+            item1dict[set_attname] = getattr(item1, attname)
+            item2dict[set_attname] = getattr(item1, attname)
+            item3dict[set_attname] = getattr(item1, attname)
+        item2dict['id'] = 10000
+        item3dict['id'] = 10001
+        item1dict["question"] = "test"
+        response = self.client.put(reverse_lazy('assessmentitem-list'), json.dumps([item1dict, item2dict, item3dict]), content_type="application/json")
+        self.assertEqual(response.status_code, 400)
+
+    def test_bulk_update_checksum_file_not_associated_create_new_file_object(self):
+        exercise = ContentNode.objects.filter(kind=content_kinds.EXERCISE).first()
+        item1 = AssessmentItem.objects.create(contentnode=exercise)
+        item1dict = {}
+        for field in AssessmentItem._meta.fields:
+            attname = field.attname
+            set_attname = attname
+            if attname == 'contentnode_id':
+                set_attname = 'contentnode'
+            item1dict[set_attname] = getattr(item1, attname)
+        checksum = "b6d83d66859b0cf095ef81120ef98e1f"
+        item1dict["question"] = "![I'm an image!]($" + exercises.IMG_PLACEHOLDER + "/{checksum}.gif)".format(checksum=checksum)
+        File.objects.create(checksum=checksum)
+        self.client.put(reverse_lazy('assessmentitem-list'), json.dumps([item1dict]), content_type="application/json")
+        self.assertEqual(File.objects.filter(checksum=checksum).count(), 2)
+
+    def test_bulk_update_checksum_file_associated_use_existing_file_object(self):
+        exercise = ContentNode.objects.filter(kind=content_kinds.EXERCISE).first()
+        item1 = AssessmentItem.objects.create(contentnode=exercise)
+        item1dict = {}
+        for field in AssessmentItem._meta.fields:
+            attname = field.attname
+            set_attname = attname
+            if attname == 'contentnode_id':
+                set_attname = 'contentnode'
+            item1dict[set_attname] = getattr(item1, attname)
+        checksum = "b6d83d66859b0cf095ef81120ef98e1f"
+        item1dict["question"] = "![I'm an image!]($" + exercises.IMG_PLACEHOLDER + "/{checksum}.gif)".format(checksum=checksum)
+        File.objects.create(checksum=checksum, assessment_item=item1)
+        self.client.put(reverse_lazy('assessmentitem-list'), json.dumps([item1dict]), content_type="application/json")
+        self.assertEqual(File.objects.filter(checksum=checksum).count(), 1)

--- a/contentcuration/contentcuration/views/files.py
+++ b/contentcuration/contentcuration/views/files.py
@@ -221,12 +221,14 @@ def exercise_image_upload(request):
         return HttpResponseBadRequest("Only POST requests are allowed on this endpoint.")
 
     fobj = request.FILES.values()[0]
+    assessment_item_id = request.POST.get('assessment_item_id', None)
     name, ext = os.path.splitext(fobj._name)
     get_hash(DjFile(fobj))
     file_object = File(
         preset_id=format_presets.EXERCISE_IMAGE,
         file_on_disk=DjFile(request.FILES.values()[0]),
         file_format_id=ext[1:].lower(),
+        assessment_item_id=assessment_item_id,
     )
     file_object.save()
     return HttpResponse(json.dumps({

